### ISSUE TITLE
Can DifferentiationInterface.jl replace the AD backend extensions here?

### ISSUE THREAD

**Issue by gdalle** [tpapp/LogDensityProblemsAD.jl#26]

**Comment by tpapp** [comment]
Yes, I think it would be great to *add* API for that backend too, but I would be reluctant to *replace* existing ones since there may be user code out there using it.

**Comment by gdalle** [comment]
I'm not sure my question was clear, sorry.
DifferentiationInterface.jl is not really a backend, it's a door to a dozen backends. In that way, it contains much of the code that is in your package extensions: take a look at
- https://github.com/tpapp/LogDensityProblemsAD.jl/blob/master/ext/LogDensityProblemsADForwardDiffExt.jl (similar to https://github.com/gdalle/DifferentiationInterface.jl/blob/main/ext/DifferentiationInterfaceForwardDiffExt/allocating.jl, with config setup)
- https://github.com/tpapp/LogDensityProblemsAD.jl/blob/master/ext/LogDensityProblemsADReverseDiffExt.jl (similar to https://github.com/gdalle/DifferentiationInterface.jl/blob/main/ext/DifferentiationInterfaceReverseDiffExt/allocating.jl, with tape setup)

The general idea is that many package developers have had to write such glue code, and if we put it in a single place it will be easier to maintain and optimize

**Comment by gdalle** [comment]
This would also allow a clean switch from symbol-based to ADTypes.jl-based backend specification, although symbols can still be supported if necessary.

Related:

- https://github.com/tpapp/LogDensityProblemsAD.jl/issues/15

**Comment by devmotion** [comment]
> This would also allow a clean switch from symbol-based to ADTypes.jl-based backend specification

LogDensityProblemsAD already supports ADTypes (this functionality is eg used by Turing).

**Comment by gdalle** [comment]
True, what I meant is that it supports ADTypes by falling back on symbols

https://github.com/tpapp/LogDensityProblemsAD.jl/blob/master/ext/LogDensityProblemsADADTypesExt.jl

If you wanna give a shot to DifferentiationInterface, it would probably become the other way around, symbols being translated into ADTypes

**Comment by tpapp** [comment]
@gdalle: Again, I am happy to expose unifying interfaces in this package. We can even use it as a fallback under the hood by translating ADTypes.

@devmotion: We probably should replace the examples in the README and the manual with ADTypes at some point.

**Comment by gdalle** [comment]
If you want I can open a demo PR changing one backend, so that you see what it would look like? E.g. ForwardDiff?

**Comment by tpapp** [comment]
Yes, PRs are always welcome.

**Comment by wsmoses** [comment]
For the reason specified in my comment https://github.com/tpapp/LogDensityProblemsAD.jl/pull/29#pullrequestreview-2051253729 this would present issues for Enzyme users until DI supports multiple arguments and activity states.
